src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -848,679 +590,376 @@ void StatefulWriter::send_heartbeat_to_all_readers()
+            if (!remoteReader->is_reliable())
+            {
+                remoteReader->acked_changes_set(change->sequenceNumber + 1);
+            }
+            else
+            {
+                remoteReader->from_unsent_to_status(
+                    change->sequenceNumber,
+                    UNACKNOWLEDGED,
+                    false);
+            }
+        }
     }
+}
 
-    try
-    {
-        uint32_t lastBytesProcessed = 0;
+DeliveryRetCode StatefulWriter::deliver_sample_to_network(
+        CacheChange_t* change,
+        RTPSMessageGroup& group,
+        LocatorSelectorSender& locator_selector,
+        const std::chrono::time_point<std::chrono::steady_clock>& max_blocking_time)
+{
+    NetworkFactory& network = mp_RTPSParticipant->network_factory();
+    DeliveryRetCode ret_code = DeliveryRetCode::DELIVERED;
+    uint32_t n_fragments = change->getFragmentCount();
+    FragmentNumber_t min_unsent_fragment = 0;
+    bool need_reactivate_periodic_heartbeat = false;
+
+    while (DeliveryRetCode::DELIVERED == ret_code &&
+            min_unsent_fragment != n_fragments + 1)
+    {
+        locator_selector.locator_selector.reset(false);
+        auto first_relevant_reader = matched_remote_readers_.begin();
+        bool inline_qos = false;
+        bool should_be_sent = false;
+        min_unsent_fragment = n_fragments + 1;
+
+        for (auto remote_reader = first_relevant_reader; remote_reader != matched_remote_readers_.end();
+                ++remote_reader)
+        {
+            SequenceNumber_t gap_seq;
+            FragmentNumber_t next_unsent_frag = 0;
+            if ((*remote_reader)->change_is_unsent(change->sequenceNumber, next_unsent_frag, gap_seq,
+                    need_reactivate_periodic_heartbeat) &&
+                    (0 == n_fragments || min_unsent_fragment >= next_unsent_frag))
+            {
+                if (min_unsent_fragment > next_unsent_frag)
+                {
+                    locator_selector.locator_selector.reset(false);
+                    first_relevant_reader = remote_reader;
+                    min_unsent_fragment = next_unsent_frag;
+                }
 
-        while (!relevantChanges.empty())
-        {
-            RTPSWriterCollector<ReaderProxy*>::Item changeToSend = relevantChanges.pop();
-            bool expectsInlineQos = false;
-            locator_selector_.reset(false);
+                (*remote_reader)->active(true);
+                locator_selector.locator_selector.enable((*remote_reader)->guid());
+                should_be_sent = true;
+                inline_qos |= (*remote_reader)->expects_inline_qos();
 
-            for (const ReaderProxy* remoteReader : changeToSend.remoteReaders)
-            {
-                locator_selector_.enable(remoteReader->guid());
-                expectsInlineQos |= remoteReader->expects_inline_qos();
+                // TODO Comprobar si el hueco esta en el history, y enviar uno gap a todos.
```

Comment:
Removed, Currently cannot be done without making a search in history and I think this will decrease performance. In the future, if there is a quick access from one CacheChange_t to the previous one. This could be done using linked list like FlowController stores the CacheChange_ts